File: extension/src/interfaces.ts
/**
 * Interfaces
 */

export enum PackageType {
    VoidResponse = 'VoidResponse',
    ErrorResponse = 'ErrorResponse',

    Hello = 'Hello',
    FindRepoRequest = 'FindRepoRequest',
    LocalRepo = 'LocalRepo',
    CloneRequest = 'CloneRequest',
    CloneResponse = 'CloneResponse',
    OpenRequest = 'OpenRequest'
}

export interface PackageBasis {
    kind: PackageType;
    uuid: string;
    // tslint:disable-next-line:no-any
    data: any;
}

export interface ErrorResponse {
    msg: string;
}

export interface HelloPackage {
    uuid: string;
    version: number;
}

export interface FindRepoRequest {
    repoName: string;
}

export interface CloneRequest {
    repoUrl: string;
    localPath?: string;
}

export interface CloneResponse {
    output: string;
}

export interface OpenRequest {
    repo: LocalRepo;
}

//

export interface LocalRepo {
    remoteRepositiory: string;
    path: string;
}

export interface Service {
    findRepo(name: string): Promise<LocalRepo>;

    clone(url: string, localPath?: string): Promise<string>;

    open(repo: LocalRepo): Promise<void>;
}

File: daemon/interfaces.go
package main

type packeageType string
const (
    // requests
    helloType           packeageType = "Hello"
    findRepoRequestType packeageType = "FindRepoRequest"
    cloneRequestType    packeageType = "CloneRequest"
    // responses
    cloneResponseType   packeageType = "CloneResponse"
    errorResponseType   packeageType = "ErrorResponse"
    voidResponseType    packeageType = "VoidResponse"
)

type packageBasis struct {
    Kind packeageType  `json:"kind"`
    UUID string        `json:"uuid"`
    Data interface{}   `json:"data"`
}

type helloPackage struct {
    UUID string   `json:"uuid"`
    Version int   `json:"version"`
}

type findRepoRequest struct {
    RepoName string  `json:"repoName"`
}

type cloneRequest struct {
    RepoUrl   string  `json:"repoUrl"`
    LocalPath string  `json:"localPath"`
}

type cloneResponse struct {
    Output string  `json:"output"`
}

type openRequest struct {
    Repo localRepo  `json:"repo"`
}

type errorResponse struct {
    Msg string  `json:"msg"`
}

//

type localRepo struct {
    RemoteRepositiory string  `json:"remoteRepositiory"`
    Path              string  `json:"path"`
}